solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,316 @@
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyGuard {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        require(
+            pendingOptimisticMints[depositKey] > 0,
+            "Optimistic minting not requested of already finalized"
+        );
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingCancelled(
+            msg.sender,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    // TODO: Find a convenient way for a guard to block minting at 3AM and deal
+    //       with an errant minter.
+
+    /// @notice Adds the address to the minter set.
+    function addMinter(address minter) external onlyOwner {
+        require(!isMinter[minter], "This address is already a minter");
+        isMinter[minter] = true;
+        emit MinterAdded(minter);
+    }
+
+    /// @notice Removes the address from the minter set.
+    function removeMinter(address minter) external onlyOwner {
+        require(isMinter[minter], "This address is not a minter");
+        delete isMinter[minter];
+        emit MinterRemoved(minter);
+    }
+
+    /// @notice Adds the address to the guard set.
+    function addGuard(address guard) external onlyOwner {
+        require(!isGuard[guard], "This address is already a guard");
+        isGuard[guard] = true;
+        emit GuardAdded(guard);
+    }
+
+    /// @notice Removes the address from the guard set.
+    function removeGuard(address guard) external onlyOwner {
+        require(isGuard[guard], "This address is not a guard");
+        delete isGuard[guard];
+        emit GuardRemoved(guard);
+    }
+
+    /// @notice Calculates deposit key the same way as the Bridge contract.
+    ///         The deposit key is computed as
+    ///         keccak256(fundingTxHash | fundingOutputIndex).
+    function calculateDepositKey(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) public view returns (uint256) {
+        return
+            uint256(
+                keccak256(abi.encodePacked(fundingTxHash, fundingOutputIndex))
+            );
+    }
+
+    /// @notice Used by TBTCVault.receiveBalanceIncrease to repay the optimistic
+    ///         minting debt before TBTC is minted. When optimistic minting is
+    ///         finalized, debt equal to the value of the deposit being
+    ///         a subject of the optimistic minting is incurred. When TBTCVault
+    ///         sweeps a deposit, the debt is fully or partially paid off, no
+    ///         matter if that particular deposit was used for the optimistic
+    ///         minting or not.
+    function repayOptimisticMintDebt(address depositor, uint256 amount)
+        internal
+        returns (uint256)
+    {
+        uint256 debt = optimisticMintingDebt[depositor];
+
+        if (amount > debt) {
```

Comment:
```
        // TODO: emit an event

        if (amount > debt) {
            optimisticMintingDebt[depositor] = 0;
            return amount - debt;
        }

        optimisticMintingDebt[depositor] -= amount;
        return 0;
```
?

Reply:
@nkuba @lukasz-zimnoch @mhluongo Thoughts on this one? The current version feels more intuitive to me.

Reply:
I prefer the current version